src/crypto/tls/ech.go

```diff
@@ -27,164 +27,160 @@ const (
+	if err := c.echUpdateClientHelloOuter(hello, helloInner, enc); err != nil {
+		return nil, nil, err
+	}
+
+	// Offer ECH.
+	c.ech.offered = true
+	helloInner.raw = nil
+	hello.raw = nil
+	return hello, helloInner, nil
+}
+
+// echUpdateClientHelloOuter is called by the client to construct the payload of
+// the ECH extension in the outer handshake.
+func (c *Conn) echUpdateClientHelloOuter(hello, helloInner *clientHelloMsg, enc []byte) error {
+	var (
+		ech echClientOuter
+		err error
+	)
+
+	// Copy all compressed extensions from ClientHelloInner into
+	// ClientHelloOuter.
+	for _, ext := range echOuterExtensions {
+		echCopyExtensionFromClientHelloInner(hello, helloInner, ext)
+	}
+
+	// Always copy the "key_shares" extension from ClientHelloInner, regardless
+	// of whether it gets compressed.
+	hello.keyShares = helloInner.keyShares
 
-	// ClientECH, the payload of the "encrypted_client_hello" extension.
-	var ech echClientOuter
 	_, kdf, aead := c.ech.sealer.Suite().Params()
 	ech.handle.suite.kdfId = uint16(kdf)
 	ech.handle.suite.aeadId = uint16(aead)
-	ech.handle.configId = echConfig.configId
-	ech.handle.enc = enc // Empty after HRR
+	ech.handle.configId = c.ech.configId
+	ech.handle.enc = enc
+
+	// EncodedClientHelloInner
+	helloInner.raw = nil
+	encodedHelloInner := echEncodeClientHelloInner(helloInner.marshal())
+	if encodedHelloInner == nil {
+		return errors.New("tls: ech: encoding of EncodedClientHelloInner failed")
+	}
 
 	// ClientHelloOuterAAD
+	hello.raw = nil
 	hello.ech = ech.marshal()
 	helloOuterAad := echEncodeClientHelloOuterAAD(hello.marshal(),
 		int(aead.CipherLen(uint(len(encodedHelloInner)))))
 	if helloOuterAad == nil {
-		return nil, nil, errors.New("tls: ech: encoding of ClientHelloOuterAAD failed")
+		return errors.New("tls: ech: encoding of ClientHelloOuterAAD failed")
 	}
 
 	ech.payload, err = c.ech.sealer.Seal(encodedHelloInner, helloOuterAad)
 	if err != nil {
-		return nil, nil, fmt.Errorf("tls: ech: seal failed: %s", err)
+		return fmt.Errorf("tls: ech: seal failed: %s", err)
 	}
 	if testingECHTriggerPayloadDecryptError {
 		ech.payload[0] ^= 0xff // Inauthentic ciphertext
 	}
 	ech.raw = nil
 	hello.ech = ech.marshal()
 
-	// Offer ECH.
-	c.ech.offered = true
 	helloInner.raw = nil
 	hello.raw = nil
-	return hello, helloInner, nil
+	return nil
 }
 
-func (c *Conn) echAcceptOrReject(hello *clientHelloMsg) (*clientHelloMsg, error) {
+// echAcceptOrReject is called by the client-facing server to determine whether
+// ECH was offered by the client, and if so, whether to accept or reject. The
+// return value is the ClientHello that will be used for the connection.
+//
+// This function is called prior to processing the ClientHello. In case of
+// HelloRetryRequest, it is also called before processing the second
+// ClientHello. This is indicated by the afterHRR flag.
+func (c *Conn) echAcceptOrReject(hello *clientHelloMsg, afterHRR bool) (*clientHelloMsg, error) {
```

Comment:
```suggestion
func (c *Conn) echAcceptOrRejectAfterHRR(hello *clientHelloMsg, afterHRR bool) (*clientHelloMsg, error) {
```
(To avoid the comments when invoking this function, i.e., `echAcceptOrRejectAfterHRR(..., false /* afterHRR */)`

Reply:
This function is called before and after HRR (hence the `afterHRR` flag).

Reply:
Yeah, I get the purpose of the flag, it just seems like `echAcceptOrRejectAfterHRR` is a more descriptive name. If I see this:

```
echAcceptOrRejectAfterHRR(hello, false)
```

I can reasonably assume that I'm processing `hello` before HRR. But if I see this:

```
echAcceptOrReject(hello, false)
```

It's less clear what `false` means, which is why the comments `/* afterHRR */` are included. 🤷

Reply:
The convention in this library is to pass a flag and not try to name the function in a way that make it clear in context what the flag does. I'll stick with that convention.